### DISCUSSION TITLE
Idea to remove trailing slashes from toplinks? or all generated links/uris? by a config setting?

### DISCUSSION THREAD

**Discussion by seansan** [OpenMage/magento-lts#2128] [Ideas]
Smally. We are seeing ocassional reports about duplicate links and google seems to see pages like `/service` and `/service/` as 2 separate pages. In addition we have an nginx rules that rewrites all uri's ending with a slash to the uri without the slash

So in essence you could say we are a "trailing slashless website'

I am not sure if this goes for all websites, maybe others DO add slashes on every page?

So, I have added a small change below to toplinks (where it goes wrong most of the time) to remove a trailing slash

In addition one could make this a new config option "Remove trailing slashes on all uris" yes/no, default to no

Where yes make an active effort to remove any trailing slashes on the final uri

https://webmasters.googleblog.com/2010/04/to-slash-or-not-to-slash.html?m=1
https://www.searchenginejournal.com/google-duplicate-content/228513/#ok

### Description (*)
See changes to method below
Might also add new config option to make it configurable (and does not break anything)

### Expected behavior (*)
Settings / ability to remove all trailing slashes

### Benefits
Better for SEO

### Possible solution

Seems to work have not found any problems with it where uris are expecting a trainling slash

change line

` 'url'           => rtrim(($prepare ? $this->getUrl($url, (is_array($urlParams) ? $urlParams : array())) : $url), '/'),`

in class Mage_Page_Block_Template_Links

```
    public function addLink($label, $url='', $title='', $prepare=false, $urlParams=array(),
        $position=null, $liParams=null, $aParams=null, $beforeText='', $afterText='')
    {
        if (is_null($label) || false===$label) {
            return $this;
        }
        $link = new Varien_Object(array(
            'label'         => $label,
            'url'           => rtrim(($prepare ? $this->getUrl($url, (is_array($urlParams) ? $urlParams : array())) : $url), '/'),
            'title'         => $title,
            'li_params'     => $this->_prepareParams($liParams),
            'a_params'      => $this->_prepareParams($aParams),
            'before_text'   => $beforeText,
            'after_text'    => $afterText,
        ));

        $this->_addIntoPosition($link, $position);

        return $this;
    }
```

### References

Here's an old video by Matt Cutts that gives a pretty good impression as to how Google appraoches things
 
https://www.youtube.com/watch?v=CTrdP7lJ2HU 
 
(note: this is an older video on that mentions page rank .. but it does illustrate the approach)
 
Like this. too
https://webmasters.googleblog.com/2010/04/to-slash-or-not-to-slash.html?m=1 
 
Finally, and more recently, here is a discussed on how John Mueller from Google explains things
https://www.seroundtable.com/google-trailing-slashes-url-24943.html

**Comment by seansan** [comment]
Idea needs a review

**Comment by m-overlund** [comment]
Wouldn't it be better to implement canonical url tags on CMS pages in general?
With some way to set preference for trailing/non-trailing slashes?

I do not think you can solve the issue completely without canonical tags - backlinks etc. can still link "incorrectly" which you have no control over.

**Comment by seansan** [comment]
I think canonicals already exist

So this is true

But the suggestion is to also fix the source link generated by Magento

External links we have no control over

But the external link / party may copy the wrong url from Magento (Meaning
that of you use wrong links yourself then the chance is higher that someone
else will use that wrong link too)

**Comment by joshua-bn** [comment]
Canonical tags are just a hint. Google will review the contents of each page and determine if they are the same. You can see this in the search console. Google won't penalize you for something like having a trailing slash. So, not really much in the way of SEO here when it comes to Google. It might make a difference for other search engines. It certainly doesn't hurt anything. I think it's a good idea.
